inc/saimpls.h

```diff
@@ -81,28 +81,15 @@ typedef enum _sai_inseg_entry_attr_t
      */
     SAI_INSEG_ENTRY_ATTR_NEXT_HOP_ID,
 
-    /**
-     * @brief Use TC for a label.
-     *
-     * Associate TC using SAI_SWITCH_ATTR_QOS_EXP_TO_TC_MAP or SAI_PORT_ATTR_QOS_EXP_TO_TC_MAP.
-     * Mainly used for E-LSP kind of MPLS tunnels, where EXP infers both TC and COLOR
-     *
-     * @type bool
-     * @flags CREATE_AND_SET
-     * @default true
-     */
-    SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP,
-
     /**
      * @brief TC for a label.
      *
-     * Associate TC by a label.
-     * Mainly used for L-LSP kind of MPLS tunnels, where label infers TC and EXP infers COLOR
+     * Associate TC by a label (override TC provided by QOS MAP)
+     * Mainly used for L-LSP tunnels, where label infers TC and EXP infers COLOR
      *
      * @type sai_uint8_t
      * @flags CREATE_AND_SET
      * @default 0
-     * @validonly SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP == false
```

Comment:
@validonly SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP == false is removed, how to identify L-LSP now?

Reply:
> @validonly SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP == false is removed, how to identify L-LSP now?

If an application passes SAI_INSEG_ENTRY_ATTR_QOS_TC then it is L-LSP. It is not mandatory argument, so if an application skips it, then it is E-LSP. @marian-pritsak recommended getting rid of validonly since its possible to achieve it without this.

Reply:
@marian-pritsak should we add QOS map object id here along with this flag? this is to enable different QoS treatment for different labels but passing through same port? Basically two tunnels may terminate through same port but want to use different QOS maps. Thoughts? @Haidong988 provided this input

Reply:
This is for MPLS VPN situation, so I suggest 

> > @validonly SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP == false is removed, how to identify L-LSP now?
> 
> If an application passes SAI_INSEG_ENTRY_ATTR_QOS_TC then it is L-LSP. It is not mandatory argument, so if an application skips it, then it is E-LSP. @marian-pritsak recommended getting rid of validonly since its possible to achieve it without this.

So if you update E-LSP into L-LSP, how could you update it back into E-LSP?

Reply:
> @marian-pritsak should we add QOS map object id here along with this flag? this is to enable different QoS treatment for different labels but passing through same port? Basically two tunnels may terminate through same port but want to use different QOS maps. Thoughts? @Haidong988 provided this input

I think it's better to put it into tunnel attribute since MPLS insegment can get tunnel attribute in MPLS VPN case. It can also be used by tunnel terminate entry in VXLAN case.

Reply:
> This is for MPLS VPN situation, so I suggest
> 
> > > @validonly SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP == false is removed, how to identify L-LSP now?
> > 
> > 
> > If an application passes SAI_INSEG_ENTRY_ATTR_QOS_TC then it is L-LSP. It is not mandatory argument, so if an application skips it, then it is E-LSP. @marian-pritsak recommended getting rid of validonly since its possible to achieve it without this.
> 
> So if you update E-LSP into L-LSP, how could you update it back into E-LSP?

I guess its a good question, on such a case, one has to either remove entry and add new one. @marian-pritsak  any thoughts here?

Reply:
> > @marian-pritsak should we add QOS map object id here along with this flag? this is to enable different QoS treatment for different labels but passing through same port? Basically two tunnels may terminate through same port but want to use different QOS maps. Thoughts? @Haidong988 provided this input
> 
> I think it's better to put it into tunnel attribute since MPLS insegment can get tunnel attribute in MPLS VPN case. It can also be used by tunnel terminate entry in VXLAN case.

there is a thought of getting rid of tunnel object altogether and use next hop attributes, there is a thought that traffic class or qos map should be part of next hop instead of tunnel. tunnel PR is separate, we have to make such changes in that PR if we want to put QoS map in tunnel

Reply:
In basic L3 case, QoS map is by port or global.
In traditional tunnel case, QoS map is by tunnel. It also works if you put it  in nexthop because nexthop means a logicport. 
In overlay tunnel with both L2 and L3 case,  QoS map is by tunnel. If you put it in nexthop, in inner l3 routing case, nexthop consist of inner layer3 edit and outer tunnel edit（DVP）. Then there could be nexthop update issue. 
The requirement extending QoS Map is caused by tunnel application actually. It's the nature model.

Reply:
> > This is for MPLS VPN situation, so I suggest
> > > > @validonly SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP == false is removed, how to identify L-LSP now?
> > > 
> > > 
> > > If an application passes SAI_INSEG_ENTRY_ATTR_QOS_TC then it is L-LSP. It is not mandatory argument, so if an application skips it, then it is E-LSP. @marian-pritsak recommended getting rid of validonly since its possible to achieve it without this.
> > 
> > 
> > So if you update E-LSP into L-LSP, how could you update it back into E-LSP?
> 
> I guess its a good question, on such a case, one has to either remove entry and add new one. @marian-pritsak any thoughts here?

In my understanding, default value 0 is a valid value. If the application created without this attribute, I believe the value should be 0 otherwise no default value is needed.
Even SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP is removed, it still exists under SAI layer （in adapt layer）. I think SAI_NEXT_HOP_ATTR_TUNNEL_VNI description need no default value, because it is needed for Inner layer3 case only. In inner L2 case, VNI is mapped from VSI. And no requirement updating a L3 nexthop into a L2 one exists I guess.

Reply:
So I am not disagreeing here, MPLS tunnel attributes are part of separate PR and we can put QoS map object id as attribute for MPLS tunnel. This PR is only for QoS map which will be applicable to LSR or mid point MPLS switches.

So here is my initial idea and in the increasing order of preference 
- QoS map at switch level
- QoS map at InSeg entry level (for tunnel termination)
- QoS map at port level

About valid-only flag, I am open to introduce it back. 

However based on earlier conversation I understood an application can skip passing it to SAI and in that case it will be E-LSP (default value of 0 even if valid won't be used because application didn't pass it), if application passes it (with or without value of 0) it will be L-LSP. But I don't know if we can switch back and forth between E-LSP and L-LSP in this case without removing entry. Not a big deal for me, I am okay if application has to remove E-LSP entry and add L-LSP entry for same label.

Reply:
> So I am not disagreeing here, MPLS tunnel attributes are part of separate PR and we can put QoS map object id as attribute for MPLS tunnel. This PR is only for QoS map which will be applicable to LSR or mid point MPLS switches.
> 
> So here is my initial idea and in the increasing order of priority for underlying switch
> 
> * QoS map at switch level
> * QoS map at port level - overrides above
> * QoS map at InSeg entry level (for tunnel termination) - overrides above
> 
> About valid-only flag, I am open to introduce it back.
> 
> However based on earlier conversation I understood an application can skip passing it to SAI and in that case it will be E-LSP (default value of 0 even if valid won't be used because application didn't pass it), if application passes it (with or without value of 0) it will be L-LSP. But I don't know if we can switch back and forth between E-LSP and L-LSP in this case without removing entry. Not a big deal for me, I am okay if application has to remove E-LSP entry and add L-LSP entry for same label.

I am also okay if application has to remove E-LSP entry and add L-LSP entry for same label.
I mean the action of Create/Get/Set API should be clear for the attribute. Under what conditions it chould be updated(Set API)? What return value should be for "Get API" for E-LSP? How cloud "Get API" know MPLS Insegment is ELSP or LLSP after MPLS insegment created?

Reply:
Hi Haidong, I have applied your comments.